From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <FinlayOxby@outlook.com>
Date: Tue, 18 Jul 2023 15:04:12 +0100
Subject: [PATCH] Graphite Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 615c47f069d429e14993c60253f719b514849a67..5e242de3e52b3ac297f9b34c223ec1dd47e9b2a6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2092,7 +2092,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Purpur end
-
+        // Graphite start - Graphite configuration
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getGraphiteConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Graphite end - Graphite configuration
         /**
          * Sends the component to the player
          *
