From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <FinlayOxby@outlook.com>
Date: Tue, 18 Jul 2023 15:04:12 +0100
Subject: [PATCH] Graphite Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index aaef58468a3c31f35e5067ed4263e9dd3fbddddd..c9fc7b84c9bc03caa01a5ce22d785f766c2022ba 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2132,7 +2132,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Purpur end
-
+        // Graphite start - Graphite configuration
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getGraphiteConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Graphite end - Graphite configuration
         /**
          * Sends the component to the player
          *
