From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <FinlayOxby@outlook.com>
Date: Sat, 29 Jul 2023 22:54:20 +0100
Subject: [PATCH] Iron Golems Attack Creepers

Iron Golems can now attack Creepers

diff --git a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
index 64aba511e615983988cdb6a0fd45b7d9d4f2f16d..f38e4fdb7d1ce53856f574005d38a1f412300950 100644
--- a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
@@ -125,7 +125,13 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         this.targetSelector.addGoal(2, new HurtByTargetGoal(this, new Class[0]));
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Mob.class, 5, false, false, (entityliving) -> {
-            return entityliving instanceof Enemy && !(entityliving instanceof Creeper);
+            // Graphite start - Iron Golems Attack Creepers
+            if (this.level().graphiteConfiguration().mobSettings.ironGolemOptions.attackCreepers) {
+                return entityliving instanceof Enemy;
+            } else {
+                return entityliving instanceof Enemy && !(entityliving instanceof Creeper);
+            }
+            // Graphite end - Iron Golems Attack Creepers
         }));
         this.targetSelector.addGoal(4, new ResetUniversalAngerTargetGoal<>(this, false));
     }
@@ -147,10 +153,17 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
 
     @Override
     protected void doPush(Entity entity) {
-        if (entity instanceof Enemy && !(entity instanceof Creeper) && this.getRandom().nextInt(20) == 0) {
-            this.setTarget((LivingEntity) entity, org.bukkit.event.entity.EntityTargetLivingEntityEvent.TargetReason.COLLISION, true); // CraftBukkit - set reason
+        // Graphite start - Iron Golems Attack Creepers
+        if (this.level().graphiteConfiguration().mobSettings.ironGolemOptions.attackCreepers) {
+            if (entity instanceof Enemy && this.getRandom().nextInt(20) == 0) {
+                this.setTarget((LivingEntity) entity, org.bukkit.event.entity.EntityTargetLivingEntityEvent.TargetReason.COLLISION, true);
+            }
+        } else {
+            if (entity instanceof Enemy && !(entity instanceof Creeper) && this.getRandom().nextInt(20) == 0) {
+                this.setTarget((LivingEntity) entity, org.bukkit.event.entity.EntityTargetLivingEntityEvent.TargetReason.COLLISION, true); // CraftBukkit - set reason
+            }
         }
-
+        // Graphite end - Iron Golems Attack Creepers
         super.doPush(entity);
     }
 
@@ -178,7 +191,11 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
 
     @Override
     public boolean canAttackType(EntityType<?> type) {
-        return this.isPlayerCreated() && type == EntityType.PLAYER ? false : (type == EntityType.CREEPER ? false : super.canAttackType(type));
+        // Graphite start - Iron Golems Attack Creepers
+        if (this.isPlayerCreated() && type == EntityType.PLAYER) return false;
+        if (type == EntityType.CREEPER) return this.level().graphiteConfiguration().mobSettings.ironGolemOptions.attackCreepers;
+        return super.canAttackType(type);
+        // Graphite end - Iron Golems Attack Creepers
     }
 
     @Override
diff --git a/src/main/java/org/graphitemc/graphite/configuration/LevelConfigurations.java b/src/main/java/org/graphitemc/graphite/configuration/LevelConfigurations.java
index 8fc66aa982820b3179e1216f3176f6e8d1079a16..61b5700455933afe0ac3d5c9431bc2e978a79337 100644
--- a/src/main/java/org/graphitemc/graphite/configuration/LevelConfigurations.java
+++ b/src/main/java/org/graphitemc/graphite/configuration/LevelConfigurations.java
@@ -22,4 +22,18 @@ public class LevelConfigurations extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     public int version = CURRENT_VERSION;
 
+    public MobOptions mobSettings;
+
+    public class MobOptions extends ConfigurationPart {
+
+        public IronGolemOptions ironGolemOptions;
+
+        public class IronGolemOptions extends ConfigurationPart {
+
+            public boolean attackCreepers = false;
+
+        }
+
+    }
+
 }
