From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <FinlayOxby@outlook.com>
Date: Sat, 29 Jul 2023 23:18:40 +0100
Subject: [PATCH] Fix MC-93018

See: https://bugs.mojang.com/browse/MC-93018

diff --git a/src/main/java/dev/graphitemc/graphite/configuration/LevelConfigurations.java b/src/main/java/dev/graphitemc/graphite/configuration/LevelConfigurations.java
index d65705e77fab5e8f2e75f304b8bcba77c786c11e..bbf2465be21ac530018267047c96d3309abd15d2 100644
--- a/src/main/java/dev/graphitemc/graphite/configuration/LevelConfigurations.java
+++ b/src/main/java/dev/graphitemc/graphite/configuration/LevelConfigurations.java
@@ -41,6 +41,7 @@ public class LevelConfigurations extends ConfigurationPart {
     public class Miscellaneous extends ConfigurationPart {
 
         public boolean fixMC31819 = true;
+        public boolean fixMC93018 = true;
 
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/Wolf.java b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
index 64bceae4d06b35fcbecb0daca2496ba30e39d995..12c55b595ac8edc4fe078d7e8ebf26937771f534 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Wolf.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
@@ -651,17 +651,20 @@ public class Wolf extends TamableAnimal implements NeutralMob {
 
     @Override
     public boolean canMate(Animal other) {
-        if (other == this) {
-            return false;
-        } else if (!this.isTame()) {
-            return false;
-        } else if (!(other instanceof Wolf)) {
-            return false;
-        } else {
-            Wolf entitywolf = (Wolf) other;
+        // Graphite start - Fix MC-93018
+        if (other == this) return false;
+        if (!(other instanceof Wolf entityWolf)) return false;
 
-            return !entitywolf.isTame() ? false : (entitywolf.isInSittingPose() ? false : this.isInLove() && entitywolf.isInLove());
+        if (this.level().graphiteConfiguration().misc.fixMC93018) {
+            return entityWolf.isInSittingPose() ? false : this.isInLove() && entityWolf.isInLove();
+        } else {
+            if (!this.isTame()) {
+                return false;
+            } else {
+                return !entityWolf.isTame() ? false : (entityWolf.isInSittingPose() ? false : this.isInLove() && entityWolf.isInLove());
+            }
         }
+        // Graphite end - Fix MC-93018
     }
 
     public boolean isInterested() {
