From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <FinlayOxby@outlook.com>
Date: Tue, 5 Sep 2023 20:26:50 +0100
Subject: [PATCH] Get Random Player

Adds a method to fetch a random online player

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index dcae9fa8f318c5e85ce175106b07812a25636430..dd4a7a0864acdc739696a5e0173ad76528600d02 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3226,4 +3226,11 @@ public final class CraftServer implements Server {
         return getServer().lagging;
     }
     // Purpur end
+    // Graphite start - Get Random Player
+    public Player getRandomPlayer() {
+        if (getOnlinePlayers().isEmpty()) return null;
+
+        return getOnlinePlayers().get(new java.util.Random().nextInt(getOnlinePlayers().size()));
+    }
+    // Graphite end - Get Random Player
 }
